src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp

```diff
@@ -556,15 +556,473 @@ void runtime::gpu::CudaKernelBuilder::get_replace_slice_op(
+                    writer << "(!filter_load_in_bounds)\n";
+                    writer << "? " << make_float_i(reg_tile_size) << "\n";
+                    writer << ": ";
+                }
+                writer << "F[(c * filter_channel_size) + entry.y].f" << reg_tile_size << ";\n";
+            }
+            writer.block_end();
+            writer << "__syncthreads();\n";
+
+            // last iteration
+            writer << "#pragma unroll\n";
+            writer << "for (int i = 0; i < NUM_ROWS; i++)\n";
+            writer.block_begin();
+            {
+                writer << "Matrix row;\n";
+                writer << "Matrix col;\n";
+                writer << "row.f" << reg_tile_size << " = a_tile[i][threadIdx.x].f" << reg_tile_size
+                       << ";\n";
+                writer << "col.f" << reg_tile_size << " = b_tile[i][threadIdx.y].f" << reg_tile_size
+                       << ";\n";
+                // accumulate the product
+                writer << "#pragma unroll\n";
+                writer << "for (int y = 0; y < " << reg_tile_size << "; y++)\n";
+                writer.block_begin();
+                {
+                    writer << "#pragma unroll\n";
+                    writer << "for (int x = 0; x < " << reg_tile_size << "; x++)\n";
+                    writer.block_begin();
+                    {
+                        writer << "result[y].f[x] += (row.f[x] * col.f[y]);\n";
+                    }
+                    writer.block_end();
+                }
+                writer.block_end();
+            }
+            writer.block_end();
+        } // end if (lookup_size_local > 0)
+        writer.block_end();
+
+        // store result block to global memory
+        writer << "int n = n_offset + threadIdx.x;\n";
+        std::string k_definition = "int k = (k_offset + threadIdx.y)";
+        std::string output_pixel = "output_pixel_idx = (output_pixel_idx * N)";
+        if (reg_tile_size == 4)
+        {
+            output_pixel += " >> 2";
+            k_definition += " << 2";
+        }
+        writer << output_pixel << ";\n";
+        writer << k_definition << ";\n";
+        writer << "if (k < K && n < N)\n";
+        writer.block_begin();
+        {
+            writer << "#pragma unroll\n";
+            writer << "for (int x = 0; x < " << reg_tile_size << "; x++)\n";
+            writer.block_begin();
+            {
+                writer << "if (k < K)\n";
+                writer.block_begin();
+                {
+                    writer << "int idx = (k * output_filter_size) + output_pixel_idx + n;\n";
+                    writer << "O[idx].f" << reg_tile_size << " = result[x].f" << reg_tile_size
+                           << ";\n";
+                }
+                writer.block_end();
+                writer << "k++;\n";
+            }
+            writer.block_end();
+        }
+        writer.block_end();
+    }
+    writer.block_end();
+}
```

Comment:
This looks good. But really complex.
One suggestion here:
Break the kernel to two: img2col for image + matrix multiplication.
The reasons are:
1. img2col is relatively simple. Even without optimization, it won't took much time.
2. matrix multiplication is already there in cuBlas and well optimized.  It take long time and need good optimization.
3. combined kernel like this are very very hard to optimize.